### PR TITLE
Add subcell boundary conditions to GRMHD

### PR DESCRIPTION
**Pull request by yoonso0-0**
## Proposed changes

Enable using subcell at external boundaries for GRMHD system.
cf. See #3746 for the same feature for Burgers.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp [hunk 1]
```diff
@@ -66,6 +76,147 @@ std::optional<std::string> Outflow::dg_outflow(
...
+        hydro::Tags::LorentzFactorTimesSpatialVelocity<DataVector, 3>;
+    using MagneticField = hydro::Tags::MagneticField<DataVector, 3>;
+    using DivergenceCleaningField =
+        hydro::Tags::DivergenceCleaningField<DataVector>;
+
+    using prim_tags_for_reconstruction =
+        tmpl::list<RestMassDensity, Pressure, LorentzFactorTimesSpatialVelocity,
+                   MagneticField, DivergenceCleaningField>;
+
+    Variables<prim_tags_for_reconstruction> boundary_vars{num_face_pts};
```
Comment (yoonso0-0) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp]: Is there a way to allocate two `Variables` (with the one on the line 197) at once to reduce overhead?

Comment (nilsdeppe) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp]: Not in one line, but my non-owning Variables PR will make this possible. It'll be 3 lines, though (probably worth it :) )

Comment (yoonso0-0) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp]: Okay, my eyes on that PR.

**Comment by yoonso0-0** [comment]
Rebased and made small fixes from Larry's review on Burgers PR.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp [hunk 2]
```diff
@@ -66,6 +78,156 @@ std::optional<std::string> Outflow::dg_outflow(
...
+    using MagneticField = hydro::Tags::MagneticField<DataVector, 3>;
+    using DivergenceCleaningField =
+        hydro::Tags::DivergenceCleaningField<DataVector>;
+
+    using prim_tags_for_reconstruction =
+        tmpl::list<RestMassDensity, Pressure, LorentzFactorTimesSpatialVelocity,
+                   MagneticField, DivergenceCleaningField>;
+
+    // Create a single large DV to reduce the number of Variables allocations
+    const size_t buffer_size_per_grid_pts =
```
Comment (yoonso0-0) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryConditions/Outflow.cpp:176]: Not sure this is the best way to count the number of individual components of `<prim_tags_for_reconstruction>`..

**Comment by nilsdeppe** [comment]
@yoonso0-0 this is still failing CI

**Comment by yoonso0-0** [comment]
I closed a separate PR for subcell FreeOutflow boundary condition, but left it as a separate commit here to make review easier. I'll squash the FD FreeOutflow commit into FD boundary condition one once review's finished.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+              ERROR("Unsupported boundary condition "
+                    << pretty_type::short_name<BoundaryCondition>()
+                    << " when using finite-difference");
+            }
+          }
+        });
+
+    // Put the computed ghost data into neighbor data with {direction,
+    // ElementId::external_boundary_id()} as the mortar_id key
+    std::vector<double> boundary_ghost_data{
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp]: Can you allocate this vector before the `Variables` and then use a non-owning Variables? That'll eliminate a copy+allocation

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        ghost_data_vars.data(),
+        ghost_data_vars.data() + ghost_data_vars.size()};
+    const std::pair mortar_id{direction, ElementId<3>::external_boundary_id()};
+
+    db::mutate<evolution::dg::subcell::Tags::NeighborDataForReconstruction<3>>(
+        box, [&mortar_id, &boundary_ghost_data](auto neighbor_data) {
+          (*neighbor_data)[mortar_id] = boundary_ghost_data;
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp]: I think here we want `std::move(boundary_ghost_data)` to eliminate another copy+allocation

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/ReconstructWork.tpp [hunk 5]
```diff
@@ -186,24 +184,36 @@ void reconstruct_prims_work(
...
+            "The neighber data is empty in direction "
+                << direction << " on element id " << element.id());
+        ghost_cell_vars[direction] = gsl::make_span(
+            &neighbor_data.at(std::pair{
+                direction,
+                *neighbors_in_direction
+                     .begin()})[vars_in_neighbor_count * neighbor_num_pts],
+            number_of_variables * neighbor_num_pts);
+      } else {
+        // retrieve boundary ghost data from neighbor_data
```
Comment (nilsdeppe) [inline src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/ReconstructWork.tpp:209]: maybe add an `ASSERT(element.external_boundaries().count(direction) == 1, "...");`?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Looks good, few minor suggestions. Please rebase and squash immediately :)

**Comment by yoonso0-0** [comment]
Squashed fixups into "Add FD boundary condition to GRMHD subcell" and rebased. I'll squash last two commits once approved by another review.

File: src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp (hunk 3 above)
Comment (yoonso0-0) [reply src/Evolution/Systems/GrMhd/ValenciaDivClean/FiniteDifference/BoundaryConditionGhostData.hpp]: Forgot to update these lines after we got the new feature :b

**Review by nilsdeppe — COMMENTED** [review]
LGTM. Given how long this PR's been open I wouldn't wait for a second reviewer. It's not required to have 2 reviewers

**Comment by yoonso0-0** [comment]
@nilsdeppe Okay!

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**